Implement `add_row_number` for Database backends, fix primary key inference for SQLite

radeusgd opened the pull request:
### Pull Request Description

Closes #6921 and also closes #7037 

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

radeusgd committed:
Fix primary key issue #7037, implement `default_ordering`

radeusgd committed:
fixes

radeusgd committed:
add missing spec

radeusgd committed:
reorder tests and add default ordering spec

radeusgd committed:
Prepare tests for DB

radeusgd committed:
Generating IR for add_row_number

radeusgd committed:
fixes

radeusgd committed:
Generate SQL code, amend tests

radeusgd committed:
doc

radeusgd committed:
changelog

GregoryTravis commented on distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:
```diff
@@ -559,10 +561,50 @@ type Table
+                problem_builder.report_other_warning (Floating_Point_Equality.Error column.name)
+        ordering = Table_Helpers.resolve_order_by self.columns order_by problem_builder
+        problem_builder.attach_problems_before on_problems <|
+            order_descriptors = case ordering.is_empty of
```
Would it be possible to call `self.order_by` and then recursively call `add_row_number`?

radeusgd replied:
That is allowed yes.

Note that
```
t1 = t0.add_row_number order_by=["X"]
t2 = t0.order_by ["X"] . add_row_number
```
has slightly different semantics.

If `t0` is sorted according to some column `Y`, then in `t1` resulting rows will still be in the `Y` order, just the added column numbering will be according to `X`; while in `t2` the rows will be re-ordered to `X`.

jdunkerley replied:
That is slightly database dependent - t1 can be reordered if no order by clause was specified.
Postgres will for example (as more efficient for it).
If multiple functions in SQL with order by clauses it tends to pick the first one I think.
Obv table order by dominates column.

radeusgd replied:
> That is slightly database dependent - t1 can be reordered if no order by clause was specified.

That's what I meant by:

> If `t0` is sorted according to some column `Y`

i.e. I assumed that `t0` has `ORDER BY Y`. And my point was that `add_row_number` itself will _not override_ this ordering, whereas an `order_by` followed by `add_row_number` will change it.

GregoryTravis commented on distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:
```diff
@@ -559,10 +561,50 @@ type Table
+            params = [SQL_Expression.Constant offset, SQL_Expression.Constant step] + order_descriptors + [separator] + grouping_expressions
+            new_expr = SQL_Expression.Operation "ROW_NUMBER" params
+
+            type_mapping = self.connection.dialect.get_type_mapping
```
This could be a helper function, I've seen it in a few places.

radeusgd replied:
Could you clarify what exactly?

I don't see what pattern you are referring to.

radeusgd committed:
Add additional wait time for cleanup of temporary tables in a test

(Sometimes the Database does cleanup slightly late and the test would fail, this should make it more stable.)

jdunkerley commented on distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:
```diff
@@ -23,6 +23,7 @@ import Standard.Table.Data.Table.Table as Materialized_Table
 import Standard.Table.Data.Type.Value_Type_Helpers
 import Standard.Table.Internal.Aggregate_Column_Helper
 import Standard.Table.Internal.Table_Helpers
+import Standard.Table.Internal.Add_Row_Number
```
Can we put in alphabetical order here please.

jdunkerley commented on distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso:
```diff
@@ -265,6 +266,21 @@ make_is_in_column arguments = case arguments.length of
+    step = arguments.at 1
+
+    separator_ix = arguments.index_of code->
+        code.build.prepare.first == "<!EXPRESSION_SEPARATOR!>"
```
Lets make this commented SQL just for the sake of safety `--<!EXPRESSION_SEPARATOR!>--`

radeusgd replied:
Ok. I also moved it into a separate variable so the constant is not duplicated.

jdunkerley reviewed (approved):
Couple of tiny comments but looks great :)

radeusgd committed:
refactor separator

radeusgd committed:
reorder imports
